esp-hal/src/lib.rs

```diff
@@ -572,6 +572,115 @@ mod critical_section_impl {
     }
 }
 
+// The state of a re-entrant lock
+pub(crate) struct LockState {
+    #[cfg(multi_core)]
+    core: portable_atomic::AtomicU8,
+}
+
+impl LockState {
+    pub const fn new() -> Self {
+        Self {
+            #[cfg(multi_core)]
+            core: portable_atomic::AtomicU8::new(0),
+        }
+    }
+
+    #[cfg(multi_core)]
+    fn current_core() -> u8 {
+        // Note: 0 means unlocked.
+        match get_core() {
+            Cpu::ProCpu => 1,
+            Cpu::AppCpu => 2,
+        }
+    }
+}
+
+// This is preferred over critical-section as this allows you to have multiple
+// locks active at the same time rather than using the global mutex that is
+// critical-section.
+#[allow(unused_variables)]
+pub(crate) fn lock<T>(state: &LockState, f: impl FnOnce() -> T) -> T {
+    // In regards to disabling interrupts, we only need to disable
+    // the interrupts that may only eb calling this function.
+
+    #[cfg(not(multi_core))]
+    {
+        // Disabling interrupts is enough on single core chips to ensure mutual
+        // exclusion.
+
+        #[cfg(riscv)]
+        return riscv::interrupt::free(f);
+        #[cfg(xtensa)]
+        return xtensa_lx::interrupt::free(|_| f());
+    }
+
+    #[cfg(multi_core)]
+    {
+        use portable_atomic::Ordering;
+
+        let current_core = LockState::current_core();
+
+        // Check for re-entry
+        {
+            // Used Relaxed ordering since an Acquire has already been performed for the
+            // re-entrant case that is being checked here.
+            let locked_core = state.core.load(Ordering::Relaxed);
+            if locked_core == current_core {
+                // Note: Since this is re-entrant, interrupts have already been disabled.
+                return f();
+            }
+        }
+
+        let mut f = Some(f);
+
+        loop {
+            // Relaxed ordering is fine here since it's just a hint for the CEX below.
+            let locked_core = state.core.load(Ordering::Relaxed);
```

Comment:
This is somewhat redundant, on the happy path, you load the core twice. I'd probably optimise this code for the happy path (performance when racing for the same lock is **probably** less frequent than the common case of just locking something on one core) important and remove this check. In a multi-core environment the other core may acquire the lock after this check anyway.

Reply:
Fair, I'll update it.

On desktop/server CPUs, it's cheaper to spin on an atomic read than to spin on an atomic write/CAS, since the CPU doesn't have to acquire an exclusive lock on the memory location.
Using an atomic read also means it won't contend with the write done for unlocking.

Though given we're only dealing with 2 cores at most (rather than the 8-64 cores that I'm used to) and contention on the same lock is highly unlikely, this "optimization" is not worth it.

Reply:
> it's cheaper to spin on an atomic read than to spin on an atomic write/CAS

I... don't know I understand why, in the end CPUs wait at the same speed. But this can also be done after we've lost the race once - lose the CAS, then spin reading, then retry the CAS after we stop spinning. Does that alternative make sense?

Reply:
Ehhh something something MESI protocol. I can't say I deeply understand it, especially in the context of ESP32s.
I suppose it's somewhat similar to using `compare_exchange_weak`.

I've opted to just remove it completely. Unless you prefer to still do it after CAS has failed. (I'm not too excited about how the code will look)

Reply:
We can come back to this if we encounter perf problems. I guess @ProfFan will point them out to us :)